tt_metal/distributed/mesh_program.hpp

```diff
@@ -0,0 +1,134 @@
+            auto devices = mesh_device->get_devices();
+            TT_ASSERT(devices.size() == this->programs.size(),
+                    "MeshProgram created for {} devices cannot be mapped to a MeshDevice with {} devices",
+                    this->programs.size(), devices.size());
+            if constexpr (std::is_same<T, void>::value) {
+                for (std::size_t program_idx = 0; program_idx < devices.size(); program_idx++) {
+                    callable(*this->programs.at(program_idx), devices.at(program_idx));
+                }
+            } else {
+                for (std::size_t program_idx = 0; program_idx < devices.size() - 1; program_idx++) {
+                    callable(*this->programs.at(program_idx), devices.at(program_idx));
+                }
+                return callable(*this->programs.at(devices.size() -1), devices.at(devices.size() -1));
+            }
+        }
+   private:
+        std::vector<std::shared_ptr<Program>> programs = {};
+
+};
+
+uint32_t CreateSemaphore(
+    MeshProgram& mesh_program,
+    const std::variant<CoreRange, CoreRangeSet> &core_spec,
+    uint32_t initial_value,
+    CoreType core_type = CoreType::WORKER);
+
+uint32_t CreateSemaphore(
+    MeshProgram& mesh_program,
+    const std::variant<CoreRange, CoreRangeSet> &core_spec,
+    uint32_t initial_value,
+    CoreType core_type,
+    chip_id_t device_id);
+
+CBHandle CreateCircularBuffer(
+    MeshProgram& mesh_program,
+    const std::variant<CoreCoord, CoreRange, CoreRangeSet> &core_spec,
+    const CircularBufferConfig &config);
+
+CBHandle CreateCircularBuffer(
+    MeshProgram& mesh_program,
+    const std::variant<CoreCoord, CoreRange, CoreRangeSet> &core_spec,
+    const CircularBufferConfig &config,
+    chip_id_t device_id);
+
+void SetRuntimeArgs(
+    MeshProgram& mesh_program,
+    KernelHandle kernel,
+    const std::variant<CoreCoord, CoreRange, CoreRangeSet> &core_spec,
+    const std::vector<uint32_t> &runtime_args);
+
+void SetRuntimeArgs(
+    MeshProgram& mesh_program,
+    KernelHandle kernel,
+    const std::variant<CoreCoord, CoreRange, CoreRangeSet> &core_spec,
+    const std::vector<uint32_t> &runtime_args,
+    chip_id_t device_id);
+
+KernelHandle CreateKernel(
+    MeshProgram& mesh_program,
+    const std::string &file_name,
+    const std::variant<CoreCoord, CoreRange, CoreRangeSet> &core_spec,
+    const std::variant<DataMovementConfig, ComputeConfig, EthernetConfig> &config);
+
+KernelHandle CreateKernel(
+    MeshProgram& mesh_program,
+    const std::string &file_name,
+    const std::variant<CoreCoord, CoreRange, CoreRangeSet> &core_spec,
+    const std::variant<DataMovementConfig, ComputeConfig, EthernetConfig> &config,
+    chip_id_t device_id);
+
+void EnqueueMeshProgram(
+    uint8_t cq_id, MeshProgram& mesh_program, std::shared_ptr<MeshDevice> mesh_device, bool blocking);
+
+void Finish(std::shared_ptr<MeshDevice> mesh_device, uint8_t cq_id);
+
```

Comment:
We should try to stick as true to the APIs as possible for now since there will be some refactor work.